### ISSUE TITLE
Function type inference

### ISSUE THREAD

**Issue by schusovskoy** [microsoft/TypeScript#38883]
**TypeScript Version:**  3.9.2

**Search Terms:**
function type

**Code**

```ts
type GetKeys = <T>(obj: T) => (keyof T)[]
const keys: GetKeys = Object.keys
```

**Expected behavior:**
The last line should throw an error, cause string[] is not assignable to (keyof T)[]

**Actual behavior:**
No error

**Playground Link:**  https://www.typescriptlang.org/play/index.html?ssl=1&ssc=1&pln=2&pc=34#code/C4TwDgpgBA4hwGkIgM5QLxQDwBUB8AFAPYBGAVgFxQ4CUGeUBA1skQGbU0DaAugFABjIgDsUwKC1RU4iZGkwB5chAHAAdJJRA

**Related Issues:** 
Maybe that https://github.com/microsoft/TypeScript/issues/38879

**Comment by MartinJohns** [comment]
This is correct. `Object.keys` returns `string[]`, **not** `(keyof T)[]`.

See #37495 (and many many many more).

**Comment by RyanCavanaugh** [comment]
https://stackoverflow.com/questions/55012174/why-doesnt-object-keys-return-a-keyof-type-in-typescript

**Comment by schusovskoy** [comment]
My question was not about why Object.keys has `(obj: object) => string[]` type, but ruther why `(obj: object) => string[]` is assignable to `<T>(obj: T) => (keyof T)`

**Comment by RyanCavanaugh** [comment]
Assignability to generic types substitutes `any` for unconstrained type parameters

**Comment by schusovskoy** [comment]
@RyanCavanaugh thanks for the explanation, but why then does the code below raise an error?

https://www.typescriptlang.org/play/index.html#code/CYUwxgNghgTiAEYD2A7AzgF3kgRgK3AwGkQBPNALngApc8q7CBKeAXgD55MYBLFAcwDaAXQBQojKQAOCAOIhiZNG3gAeACrta+KupYcaAazJIAZvD0jRydFmPkq8xeRWMwztEA

**Comment by RyanCavanaugh** [comment]
@schusovskoy the definition you have there doesn't match `Object.keys`'s definition

**Comment by schusovskoy** [comment]
If I understand correctly there are next object types:
`Object` – is any object including primitives
`{}` – is the same as `Object`
`object` – any non primitive object

Object.keys has type
```ts
type Keys = {
  (o: object): string[]
  (o: {}): string[]
}
```

I guess it's the same as
```ts
type Keys = ((o: object) => string[]) & ((o: {}) => string[])
```

I thought it means that Object.keys is the function which can accept any non primitive object and returns array of string and also it's the function which can accept any object including primitives and returns array of string. So it looks like the second case covers the first one.
I thought that `(o: {}) => string[]` is equivalent to ` ((o: object) => string[]) & ((o: {}) => string[])`. But it's also not assignable to `<T>(obj: T) => (keyof T)[]`. So only intersection of the types is assignable.

I really want to understand how and why it's work like this. So can you help me with that please.

**Comment by schusovskoy** [comment]
@MartinJohns @RyanCavanaugh Guys excuse me for my impatience but I can’t find any information, and it seems that you’re the only people who can help me.
